src/libostree/ostree-sign-ed25519.c

```diff
@@ -0,0 +1,588 @@
+    }
+
+  g_debug ("verify: data hash = 0x%x", g_bytes_hash(data));
+
+  for (gsize i = 0; i < g_variant_n_children(signatures); i++)
+    {
+      g_autoptr (GVariant) child = g_variant_get_child_value (signatures, i);
+      g_autoptr (GBytes) signature = g_variant_get_data_as_bytes(child);
+
+      g_autofree char * hex = g_malloc0 (crypto_sign_PUBLICKEYBYTES*2 + 1);
+
+      g_debug("Read signature %ld: %s", i, g_variant_print(child, TRUE));
+
+      for (GList *public_key = sign->public_keys;
+           public_key != NULL;
+           public_key = public_key->next)
+        {
+          if (crypto_sign_verify_detached ((guchar *) g_variant_get_data (child),
+                                           g_bytes_get_data (data, NULL),
+                                           g_bytes_get_size (data),
+                                           public_key->data) != 0)
+            {
+              /* Incorrect signature! */
+              g_debug("Signature couldn't be verified with key '%s'", 
+                      sodium_bin2hex (hex, crypto_sign_PUBLICKEYBYTES*2+1, public_key->data, crypto_sign_PUBLICKEYBYTES));
+            }
+          else
+            {
+              ret = TRUE;
+              g_debug ("Signature verified successfully with key '%s'",
+                       sodium_bin2hex (hex, crypto_sign_PUBLICKEYBYTES*2+1, public_key->data, crypto_sign_PUBLICKEYBYTES));
+              break;
+            }
+        }
+    }
+
+  if (ret != TRUE)
+    g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                         "Not able to verify: no valid signatures found");
+#endif /* HAVE_LIBSODIUM */
+
+  return ret;
+err:
+  return FALSE;
+}
+
+gboolean
+ostree_sign_ed25519_keypair_generate (OstreeSign *self,
+                                      GVariant **out_secret_key,
+                                      GVariant **out_public_key,
+                                      GError **error)
+ {
+  g_debug ("%s enter", __FUNCTION__);
+  g_return_val_if_fail (OSTREE_IS_SIGN (self), FALSE);
+
+  OstreeSignEd25519 *sign = ostree_sign_ed25519_get_instance_private(OSTREE_SIGN_ED25519(self));
+
+  if (sign->initialized != TRUE)
+    {
+      g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                          "Not able to sign -- libsodium library isn't initialized properly");
+      goto err;
+    }
+
+#ifdef HAVE_LIBSODIUM
+  unsigned char pk[crypto_sign_PUBLICKEYBYTES];
+  unsigned char sk[crypto_sign_SECRETKEYBYTES];
+
+  if (crypto_sign_keypair(pk, sk))
+    {
+      g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                          "Not able to generate keypair");
+      goto err;
+    }
+
+  *out_secret_key = g_variant_new_fixed_array (G_VARIANT_TYPE_BYTE, sk, crypto_sign_SECRETKEYBYTES, sizeof(guchar));
+  *out_public_key = g_variant_new_fixed_array (G_VARIANT_TYPE_BYTE, pk, crypto_sign_PUBLICKEYBYTES, sizeof(guchar));
+
+  return TRUE;
+#endif /* HAVE_LIBSODIUM */
+
+err:
```

Comment:
Useless label

Reply:
2 jumps above -- prefer to exit in that way.